compiler/src/dotty/tools/dotc/core/TypeErasure.scala

```diff
@@ -394,32 +394,92 @@ object TypeErasure {
   }
 
   /** The erased greatest lower bound of two erased type picks one of the two argument types.
-   *  It prefers, in this order:
-   *  - arrays over non-arrays
-   *  - subtypes over supertypes, unless isJava is set
-   *  - real classes over traits
+   *
+   *  This operation has the following the properties:
+   *  - Associativity and commutativity, because this method act as the minimum
+   *    of the total order induced by `compareErasedGlb`.
+   *  - Java compatibility: intersections that would be valid in Java code are
+   *    erased like javac would erase them (a Java intersection is composed of
+   *    exactly one class and one or more interfaces and always erases to the
+   *    class).
```

Comment:
In our last discussion, hadn't we come to the conclusion that we should never return an Interface as the erasure of an intersection type (unless an interface dominates everything else, of course)? IIRC the issue was that `Object & T1 & T2` would otherwise have a different erasure than `T1 & T2`, even though `Object & T1 & T2` can simplify to `T1 & T2` and though Java would erase to `Object`.

Reply:
> In our last discussion, hadn't we come to the conclusion that we should never return an Interface as the erasure of an intersection type 

Yes, I went with this version originally but decided it wasn't worth it in the end:
- Many more types end up being erased to Object which can lead to overloading clashes (it happened in the community build for example), I fear this could be a big problem in the wild making cross-compilation difficult.
- Admittedly, fact that `Object & T1 & T2` has a different erasure than `T1 & T2` isn't great, but then so is the fact
  `Object & T1` has a different erasure than `T1`.
- It turns out that intersection of traits in Java signatures actually work.
- Having erasedGlb(A, B) always return either A or B allows us to implement it as the minimum of a total order, which gives us commutativity and associativity for free and in general makes it easier to reason about what this method does.

Reply:
I see. IIUC, this is the critical aspect that allows this approach to work at all:

> It turns out that intersection of traits in Java signatures actually work.

Reply:
It's helpful, but if those Java signatures weren't allowed, we could also emit a signature containing just the first trait as an approximation, we already have to use approximated signatures in various situations where our type system is more expressive than Java's.